SST build issue on Titan

Building on Titan fails with the following error. Built as 
```
cmake -DCMAKE_INSTALL_PREFIX=$HOME/opt/adios2/installation/ -DADIOS2_USE_DataMan=OFF -DADIOS2_USE_HDF5=OFF -DADIOS2_USE_Fortran=ON -DCMAKE_BUILD_TYPE=Release ../
```

```
[ 38%] Built target sst
Scanning dependencies of target adios2
../../../lib/libadios2_enet.a(unix.c.o): In function `enet_address_set_host':
/ccs/home/kmehta/opt/adios2/thirdparty/enet/enet/unix.c:126: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
../../../lib/libadios2_evpath.a(ip_config.c.o): In function `get_qual_hostname':
/ccs/home/kmehta/opt/adios2/thirdparty/EVPath/EVPath/ip_config.c:344: warning: Using 'gethostbyaddr' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
../../../lib/libadios2_evpath.a(ip_config.c.o): In function `get_self_ip_iface':
/ccs/home/kmehta/opt/adios2/thirdparty/EVPath/EVPath/ip_config.c:110: warning: Using 'gethostbyname' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
../../../lib/libadios2_ffs.a(standard.c.o): In function `dlload_externs':
/ccs/home/kmehta/opt/adios2/thirdparty/ffs/ffs/cod/standard.c:687: undefined reference to `dlopen'
/ccs/home/kmehta/opt/adios2/thirdparty/ffs/ffs/cod/standard.c:690: undefined reference to `dlsym'
/usr/bin/ld: link errors found, deleting executable `../../../bin/cmprobe'
/usr/bin/sha1sum: ../../../bin/cmprobe: No such file or directory
collect2: error: ld returned 1 exit status
make[2]: *** [bin/cmprobe] Error 1
```

Missing `-ldl` somewhere?

--- a/thirdparty/ffs/ffs/CMakeLists.txt
+++ b/thirdparty/ffs/ffs/CMakeLists.txt
@@ -98,27 +98,28 @@ CHECK_TYPE_SIZE("long double" SIZEOF_LONG_DOUBLE)
 CHECK_TYPE_SIZE("long long" SIZEOF_LONG_LONG)
 CHECK_TYPE_SIZE("size_t" SIZEOF_SIZE_T)
 
-# use, i.e. don't skip the full RPATH for the build tree
-SET(CMAKE_SKIP_BUILD_RPATH  FALSE)
+if(BUILD_SHARED_LIBS)
+  # use, i.e. don't skip the full RPATH for the build tree
+  set(CMAKE_SKIP_BUILD_RPATH  FALSE)
 
-# when building, don't use the install RPATH already
-# (but later on when installing)
-SET(CMAKE_BUILD_WITH_INSTALL_RPATH FALSE) 
+  # when building, don't use the install RPATH already
+  # (but later on when installing)
+  set(CMAKE_BUILD_WITH_INSTALL_RPATH FALSE) 
 
-SET(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
+  set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
 
-# add the automatically determined parts of the RPATH
-# which point to directories outside the build tree to the install RPATH
-SET(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+  # add the automatically determined parts of the RPATH
+  # which point to directories outside the build tree to the install RPATH
+  set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
 
+  # the RPATH to be used when installing, but only if it's not a system directory
+  list(FIND CMAKE_PLATFORM_IMPLICIT_LINK_DIRECTORIES "${CMAKE_INSTALL_PREFIX}/lib" isSystemDir)
+  if(isSystemDir STREQUAL "-1")
+     set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
+  endif()
 
-# the RPATH to be used when installing, but only if it's not a system directory
-LIST(FIND CMAKE_PLATFORM_IMPLICIT_LINK_DIRECTORIES "${CMAKE_INSTALL_PREFIX}/lib" isSystemDir)
-IF("${isSystemDir}" STREQUAL "-1")
-   SET(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib")
-ENDIF("${isSystemDir}" STREQUAL "-1")
-
-set (CMAKE_MACOSX_RPATH 1)
+  set(CMAKE_MACOSX_RPATH 1)
+endif()
 
 if (NOT (DEFINED NO_SOCKETS))
     check_function_exists(socket SOCKETS_FOUND)
@@ -204,8 +205,11 @@ if(UNIX)
     list(APPEND _pkg_config_private_libs m)
 endif()
 if(SHARED_LIBS_SUPPORTED)
+    unset(NO_DYNAMIC_LINKING)
     target_link_libraries(ffs PRIVATE ${CMAKE_DL_LIBS})	
-    list(APPEND _pkg_config_private_libs ${CMAKE_DL_LIBS})	
+    list(APPEND _pkg_config_private_libs ${CMAKE_DL_LIBS})
+else()
+    set(NO_DYNAMIC_LINKING TRUE)
 endif()
 
 target_include_directories(ffs PUBLIC

--- a/thirdparty/ffs/ffs/cod/tests/compound_assignment.c
+++ b/thirdparty/ffs/ffs/cod/tests/compound_assignment.c
@@ -1,6 +1,6 @@
 #include "config.h"
 #include "cod.h"
-#include <strings.h>
+#include <string.h>
 #include <stdlib.h>
 #include <assert.h>
 #include <stdio.h>

--- a/thirdparty/ffs/ffs/cod/tests/t12.c
+++ b/thirdparty/ffs/ffs/cod/tests/t12.c
@@ -3,7 +3,7 @@
 #include "assert.h"
 #include <stdio.h>
 #include <stdlib.h>
-#include <strings.h>
+#include <string.h>
 #ifdef HAVE_MALLOC_H
 #include <malloc.h>
 #endif

--- a/thirdparty/ffs/ffs/cod/tests/t3.c
+++ b/thirdparty/ffs/ffs/cod/tests/t3.c
@@ -6,7 +6,7 @@
 
 #include "config.h"
 #include "cod.h"
-#include <strings.h>
+#include <string.h>
 #include <stdlib.h>
 #include <assert.h>
 #include <stdio.h>

--- a/thirdparty/ffs/ffs/cod/tests/t8.c
+++ b/thirdparty/ffs/ffs/cod/tests/t8.c
@@ -4,7 +4,7 @@
 #include "config.h"
 #include "cod.h"
 #include "assert.h"
-#include <strings.h>
+#include <string.h>
 #include <stdlib.h>
 #include <stdio.h>
 

--- a/thirdparty/ffs/ffs/config.h.cmake
+++ b/thirdparty/ffs/ffs/config.h.cmake
@@ -143,5 +143,4 @@
 #define LIBRARY_EXT "@CMAKE_SHARED_LIBRARY_SUFFIX@"
 
 /* Set to 1 to build without dynamic linking  */
-#cmakedefine NO_DYNAMIC_LINKING @SHARED_LIBS_SUPPORTED@
-
+#cmakedefine01 NO_DYNAMIC_LINKING 

--- a/thirdparty/ffs/ffs/ffs/ffs_file.c
+++ b/thirdparty/ffs/ffs/ffs/ffs_file.c
@@ -12,6 +12,7 @@
 #include "unistd.h"
 #include "errno.h"
 #include "string.h"
+#include <arpa/inet.h>
 
 #include "io_interface.h"
 

--- a/thirdparty/ffs/ffs/ffs/tests/context_test.c
+++ b/thirdparty/ffs/ffs/ffs/tests/context_test.c
@@ -17,6 +17,7 @@
 #include <sys/uio.h>
 #include <string.h>
 #include <assert.h>
+#include <arpa/inet.h>
 #include "ffs.h"
 
 #include "test_funcs.h"

--- a/thirdparty/ffs/ffs/ffs/tests/test_file_unique.c
+++ b/thirdparty/ffs/ffs/ffs/tests/test_file_unique.c
@@ -36,6 +36,7 @@ file_checksum(char *filename)
 
 int
 main(argc,argv)
+int argc;
 char **argv;
 {
     unsigned int base_sum, sum;

--- a/thirdparty/ffs/ffs/fm/fm_formats.c
+++ b/thirdparty/ffs/ffs/fm/fm_formats.c
@@ -3021,12 +3021,13 @@ int verbose;
  
     if ((iovar->string == 0) &&
 	(iovar->var_array == 0) &&
+	(index(iofield->field_type, '*') == NULL) &&
 	(iovar->data_type != unknown_type)) {
 	/* must be simple data type or array of simple data types */
 	return;
     }
     dimension = FMget_array_element_count(fmformat, iovar, data, encode);
-    if (iovar->var_array) {
+    if ((iovar->var_array) || (index(iofield->field_type, '*') != NULL)) {
 	FMgetFieldStruct descr;  /* OK */
 	long tmp_offset;
 	descr.offset = iofield->field_offset;
@@ -3064,7 +3065,7 @@ int verbose;
 	    data_offset += sub_field_size;
 	}
     }
-    if (iovar->var_array) {
+    if ((iovar->var_array) || (index(iofield->field_type, '*') != NULL)) {
 	free(data);
     }    
 }

